File: 4-envelope-encryption-and-kms/1-java/src/main/java/com/starci/envelopeapi/RecordsController.java
package com.starci.envelopeapi;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.Optional;

@RestController
public class RecordsController {

    private final SecureRecordRepository repo;
    private final EnvelopeService envelope;

    public RecordsController(SecureRecordRepository repo, EnvelopeService envelope) {
        this.repo = repo;
        this.envelope = envelope;
    }

    /** POST /records — encrypt and store ciphertext + wrapped DEK. */
    @PostMapping("/records")
    public ResponseEntity<Map<String, Object>> create(@RequestBody Map<String, String> body) throws Exception {
        Map<String, String> env = envelope.encrypt(body.get("plaintext"));
        SecureRecord rec = new SecureRecord();
        rec.setWrappedDek(env.get("wrappedDek"));
        rec.setIv(env.get("iv"));
        rec.setCiphertext(env.get("ciphertext"));
        SecureRecord saved = repo.save(rec);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(Map.of("id", saved.getId(), "status", "encrypted"));
    }

    /** GET /records/{id} — unwrap the DEK, decrypt, return the plaintext (422 if KEK shredded). */
    @GetMapping("/records/{id}")
    public ResponseEntity<Map<String, Object>> findOne(@PathVariable Integer id) {
        Optional<SecureRecord> found = repo.findById(id);
        if (found.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("status", "not_found"));
        }
        SecureRecord rec = found.get();
        try {
            String plaintext = envelope.decrypt(rec.getWrappedDek(), rec.getIv(), rec.getCiphertext());
            return ResponseEntity.ok(Map.of("id", rec.getId(), "plaintext", plaintext));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(Map.of("status", "unrecoverable"));
        }
    }

    /** POST /kek/rotate — bump the KEK version; old wrapped DEKs stay valid. */
    @PostMapping("/kek/rotate")
    public ResponseEntity<Map<String, Object>> rotate() {
        long version = envelope.rotateKek();
        return ResponseEntity.ok(Map.of("status", "rotated", "version", version));
    }

    /** POST /kek/shred — crypto-shredding: delete the KEK; all records become unrecoverable. */
    @PostMapping("/kek/shred")
    public ResponseEntity<Map<String, Object>> shred() {
        envelope.shredKek();
        return ResponseEntity.ok(Map.of("status", "shredded"));
    }
}

File: 4-envelope-encryption-and-kms/0-typescript/src/records/records.controller.ts
import {
    Body,
    Controller,
    Get,
    HttpCode,
    HttpException,
    HttpStatus,
    Logger,
    Param,
    Post,
} from "@nestjs/common"
import { InjectRepository } from "@nestjs/typeorm"
import { Repository } from "typeorm"
import { SecureRecord } from "../entities/record.entity"
import { EnvelopeService } from "../crypto/envelope.service"
import { TransitService } from "../transit/transit.service"

@Controller()
export class RecordsController {
    private readonly logger = new Logger(RecordsController.name)

    constructor(
        @InjectRepository(SecureRecord)
        private readonly repo: Repository<SecureRecord>,
        private readonly envelope: EnvelopeService,
    ) {}

    /** POST /records — encrypt the plaintext and store ciphertext + wrapped DEK. */
    @Post("records")
    @HttpCode(HttpStatus.CREATED)
    async create(@Body() body: { plaintext: string }): Promise<{ id: number; status: string }> {
        const env = await this.envelope.encrypt(body.plaintext)
        const saved = await this.repo.save(this.repo.create(env))
        this.logger.log(`Record ${saved.id} encrypted (DEK wrapped by KEK).`)
        return { id: saved.id, status: "encrypted" }
    }

    /** GET /records/:id — unwrap the DEK, decrypt, return the plaintext. */
    @Get("records/:id")
    async findOne(@Param("id") id: string): Promise<{ id: number; plaintext: string }> {
        const record = await this.repo.findOneBy({ id: parseInt(id, 10) })
        if (!record) {
            throw new HttpException({ status: "not_found" }, HttpStatus.NOT_FOUND)
        }
        try {
            const plaintext = await this.envelope.decrypt(record)
            return { id: record.id, plaintext }
        } catch {
            // KEK was shredded (or unreachable): the DEK can no longer be unwrapped.
            throw new HttpException({ status: "unrecoverable" }, HttpStatus.UNPROCESSABLE_ENTITY)
        }
    }

    /** POST /kek/rotate — rotate the KEK to a new version; old wrapped DEKs stay valid. */
    @Post("kek/rotate")
    @HttpCode(HttpStatus.OK)
    async rotate(): Promise<{ status: string; version: number }> {
        const version = await TransitService.rotateKek()
        this.logger.log(`KEK rotated to version ${version}.`)
        return { status: "rotated", version }
    }

    /** POST /kek/shred — crypto-shredding: delete the KEK; all records become unrecoverable. */
    @Post("kek/shred")
    @HttpCode(HttpStatus.OK)
    async shred(): Promise<{ status: string }> {
        await TransitService.shredKek()
        return { status: "shredded" }
    }
}
